### ISSUE TITLE
Enable mergify.io for flux-core

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#2002]
I've been experimenting with mergify.io for automatic rebase and merge of PRs. It seems to be working as designed, though I've only done minimal testing so far.

The benefit of mergify.io over other services I looked at (Including [bors-ng](https://bors.tech/)) is that mergify supports our current rebase+merge workflow with its [strict merge workflow](https://doc.mergify.io/strict-workflow.html).

Some issues with the other two solutions I looked at are:
 - bors-ng: only supports octopus merge of PRs
 - autorebase: rebases via github api which creates a lot of repo "noise" (one branch per cherry-picked commit, though the branches are eventually deleted)
 - other solutions require you to run on your own infrastructure.

I'd propose we use the following config:
```yaml
pull_request_rules:
  - name: rebase and merge when passing all checks
    conditions:
      - base=master  
      - status-success=continuous-integration/travis-ci/pr
      - status-success=codecov/project
      - label="merge-when-passing"
      - label!="work-in-progress"
      - -title~="^WIP"
      - "#changes-requested-reviews-by=0"
    actions:
      merge:
        method: merge
        strict: true
        strict_method: rebase
```

This config will cause the mergify to rebase a PR on current master when master changes, and subsquently merge the PR when all of the above conditions are true, namely:

 - target branch is master
 - Travis-CI checks pass
 - Codecov doesn't report a drop in overall project code coverage
 - A label merge-when-passing is set
 - A label work-in-progress is *not* set, and the title of the PR doesn't begin with "WIP"
 - There are no outstanding changes requested by reviewers

We could also require one or more review approval, etc. I think once a project member thinks a PR is good, then can just enable the merge-when-passing label. The rest of the checks are really there to make sure we don't accidentally merge a PR prematurely.

The drawbacks of this approach are:
 - Since the PR branch gets rebased automatically, the author may have out-of-date branch and not know it.
 - There is no chance for the person merging the PR to be able to edit the merge commit body, so we need to be cognizant that the PR title will appear directly in the merge commit (and thus our `relnotes.sh` script)

Other things to be aware of:
 - All merge commits will have "mergify[bot]" as the author (not a big deal)
 - mergify merges the first pending PR that passes checks, not the order in which PRs were submitted or had the final condition met.
 - We still have to babysit travis and restart builders on false negatives

Here's an example of a PR against my own fork being managed by mergify:

 https://github.com/grondo/flux-core/pull/7

Note that mergify rebases show up as `force-pushed by <author>`, and mergify actually supplies a good bit of information about its state  re the PR in the "checks" tab for the PR.

![mergify](https://user-images.githubusercontent.com/741970/52528040-3c477080-2c88-11e9-96f3-1b82b068ad9f.png)
j

**Comment by SteVwonder** [comment]
This is awesome.  Thanks for looking into this @grondo!

What happens when mergify goes to rebase the branch and it can't cleanly?  I presume it doesn't push anything and drops a notification in the PR?

> We could also require one or more review approval, etc.

:+1: for that suggestion

**Comment by garlick** [comment]
Looks great!  Let's give it a try!

**Comment by grondo** [comment]
> What happens when mergify goes to rebase the branch and it can't cleanly? I presume it doesn't push anything and drops a notification in the PR?

I'm not sure, I didn't try this yet. My guess is that it doesn't attempt to rebase if GitHub has a notification that the PR branch is "out of date" with master..

**Comment by garlick** [comment]
Closed by #2004

**Comment by tibdex** [comment]
Hi @grondo, I'm the author of https://github.com/tibdex/autorebase and I saw that you mentioned it here. I know that Autorebase creates Git branches while doing its thing but, as you said, they are deleted eventually. Can I then ask you why these transient branches and commits bother you? Do you know you can ignore them in CI (see https://github.com/tibdex/autorebase/issues/15)? Thanks ;)
